### PR TITLE
Tl 67 Filter items by labels, 

### PR DESCRIPTION
**Pull request by peterevilla**
[Jira story](https://triplist.atlassian.net/secure/RapidBoard.jspa?rapidView=1&modal=detail&selectedIssue=TL-67)

### What does this PR do?
Filter items by label filtering itemIds from current trip list. Implemented a debounce input, the text input is not case sensitive. 
### Screenshots
8C49082D-FC6E-432B-B33E-919195AE7FD1

### CODE REVIEW

File: src/routes/Trip/Items.tsx [hunk 1]
```diff
@@ -49,6 +51,15 @@ const Items = ({
...
 
+  let listOfItemIds = [];
+
+  const queryHandler = (value: any) => {
+    //Value from the FilterItemForm
+
+    setItemQuery(value);
+  };
+
+  const itemFilterIds: any[] = [];
```
Comment (LuisRevillaM) [inline src/routes/Trip/Items.tsx:62]: A few suggestions here: 

1. maybe we move `listOfItemIds` and `itemFilterIds` a bit closer to the filtering process?

2. Maybe we rename the empty array `itemFilterIds` to `filteredItemIds` ?
and 

3. I'm thinking we can directly assign all the list's ids to `listOfItemIds`, then the final mapping (for checked and unchecked) will only happen over `filteredItemIds`.  

```
  if (isListPresent) {
    const listOfItemIds = listsById[currentTrip.listId].itemIds;
    let filteredItemIds = [];
```

This way, if the query is empty, the `filteredItemIds` will be all of the `listOfItemIds`

```
    if (itemQuery === '') {
      //Check if form field has any value
      filteredItemIds = listOfItemIds;

    } else {
      filteredItemIds = listOfItemIds.filter((itemId:any)=>{
        return items[itemId].label.toLowerCase().includes(itemQuery.toLocaleLowerCase());
      });
    }
```

File: src/routes/Trip/Items.tsx [hunk 2]
```diff
@@ -64,9 +75,21 @@ const Items = ({
...
+      //Check if form field has any value
+      listOfItemIds = listsById[currentTrip.listId].itemIds;
+    } else {
+      const itemFilter = listsById[currentTrip.listId].items.filter((item: any) =>
+        item.label.toLowerCase().includes(itemQuery.toLowerCase()),
+      );
+      itemFilter.forEach((item: any) => {
+        return itemFilterIds.push(item.id);
+      });
+      listOfItemIds = itemFilterIds; //New List of items product of filtering
```
Comment (LuisRevillaM) [inline src/routes/Trip/Items.tsx:90]: As discussed in Slack and suggested above, we can filter the `itemIds` directly, using the `items` object to perform the comparison.

This way we save ourselves the extra loop to push the ids, and also fix the bug! 

```
      filteredItemIds = listsById[currentTrip.listId].itemIds.filter((itemId:any)=>{
        return items[itemId].label.toLowerCase().includes(itemQuery.toLocaleLowerCase());
      });
```

File: src/routes/Trip/Items.tsx [hunk 3]
```diff
@@ -77,9 +100,14 @@ const Items = ({
       },
       { checked: [], unchecked: [] },
     );
+
     return (
       <div>
         <div className="bg-indigo-200 text-center mt-16 text-xl">{listsById[listId].title}</div>
+        <FilterItemForm queryHandler={queryHandler} />
+        {itemQuery !== '' && listOfItemIds.length === 0 && (
```
Comment (LuisRevillaM) [inline src/routes/Trip/Items.tsx:108]: We would have to change this to check `filteredItemIds.length` instead

File: src/routes/Trip/Items.tsx [hunk 4]
```diff
@@ -64,9 +75,21 @@ const Items = ({
...
+    } else {
+      const itemFilter = listsById[currentTrip.listId].items.filter((item: any) =>
+        item.label.toLowerCase().includes(itemQuery.toLowerCase()),
+      );
+      itemFilter.forEach((item: any) => {
+        return itemFilterIds.push(item.id);
+      });
+      listOfItemIds = itemFilterIds; //New List of items product of filtering
+    }
+    const itemsByState = listOfItemIds.reduce(
```
Comment (LuisRevillaM) [inline src/routes/Trip/Items.tsx:92]: The reduction will now be over the `filteredItemIds`

File: yarn.lock [hunk 5]
```diff
@@ -8886,6 +8891,14 @@ react-app-polyfill@^1.0.6:
...
     whatwg-fetch "^3.0.0"
 
+react-debounce-input@^3.2.2:
+  version "3.2.2"
+  resolved "https://registry.yarnpkg.com/react-debounce-input/-/react-debounce-input-3.2.2.tgz#d2cc99c1ce47fae89037965f5699edc1b0317197"
+  integrity sha512-RIBu68Cq/gImKz/2h1cE042REDqyqj3D+7SJ3lnnIpJX0ht9D9PfH7KAnL+SgDz6hvKa9pZS2CnAxlkrLmnQlg==
+  dependencies:
+    lodash.debounce "^4"
+    prop-types "^15.7.2"
+
```
Comment (LuisRevillaM) [inline yarn.lock:8901]: Let's keep this for now, but I will create a ticket to replace it with our own debounce function.

**Comment by peterevilla** [comment]
Thanks for all the feedback!

File: src/routes/Trip/Items.tsx [hunk 6]
```diff
@@ -49,6 +51,14 @@ const Items = ({
 
   const isListPresent = currentTrip && listsById[listId];
 
+  let listOfItemIds = [];
```
Comment (LuisRevillaM) [inline src/routes/Trip/Items.tsx:54]: we can remove this declaration here and leave it as a `const` below when you do the assigning.

File: src/routes/Trip/Items.tsx [hunk 7]
```diff
@@ -66,7 +76,18 @@ const Items = ({
   }
 
   if (isListPresent) {
-    const itemsByState = listsById[currentTrip.listId].itemIds.reduce(
+    //ITEM FILTER
+    listOfItemIds = listsById[currentTrip.listId].itemIds;
```
Comment (LuisRevillaM) [inline src/routes/Trip/Items.tsx:80]: as mentioned above, we can declare and assign right here with `const`.

File: src/components/icons/deleteIcon 2.tsx [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+    <svg width="12" height="16" viewBox="0 0 14 18" fill="none" xmlns="http://www.w3.org/2000/svg">
+      <path
+        d="M9.12 7.47L7 9.59L4.87 7.47L3.46 8.88L5.59 11L3.47 13.12L4.88 14.53L7 12.41L9.12 14.53L10.53 13.12L8.41 11L10.53 8.88L9.12 7.47ZM10.5 1L9.5 0H4.5L3.5 1H0V3H14V1H10.5ZM1 16C1 17.1 1.9 18 3 18H11C12.1 18 13 17.1 13 16V4H1V16ZM3 6H11V16H3V6Z"
+        fill="gray"
+      />
+    </svg>
+  );
+};
+
+export default deleteIcon;
```
Comment (LuisRevillaM) [inline src/components/icons/deleteIcon 2.tsx:14]: This filed got copied accidentally, let's delete it!

**Pull request merged**